System test with robot framework

feerrenrut opened the pull request:
### Link to issue number:
#708

### Summary of the issue:
Introduction of a system testing framework for NVDA.

### Description of how this pull request fixes the issue:
Based on the Robot Framework, tests are written in `robot` files. The tests are discovered and run by the Framework. Each tests sets up an environment for NVDA to run in, starts NVDA as a separate process, and connects to a server running as an NVDA addon. The addon provides access to NVDA internals eg speech commands. The test code then interacts with NVDA as a user would by sending keypresses and checks that the behaviour is as expected using data from the addon.

For more information, see the readme in this PR.

### Testing performed:
Run locally against source and installed NVDA, and on appveyor

### Known issues with pull request:
Tests currently only look at simple startup/shutdown behaviour.
Developers will need to manually install dependencies via pip in order to be able to run the system tests locally.

### Change log entry:
Developers:
A system test framework has been introduced for NVDA. (#708)
A new extension point action for NVDA post startup is now available.

feerrenrut committed:
WIP prototype sysTest

Requires:
robotFramework - the system test framework
pyautogui - for controlling keyboard/mouse

To install:
pip install robotframework
pip install pyautogui

feerrenrut committed:
WIP include flaky way of exiting

Ensure this is run from a cmd prompt. Other shells may cause issues,
for instance, NVDA does not get correct focus when run from cygwin.

feerrenrut committed:
WIP System testing

Remote server implemenation and external abstracted library.

michaelDCurran committed:
Merge branch 'master' into systemTestWithRobot

michaelDCurran committed:
First try at including system tests in appveyor.

michaelDCurran committed:
Disable pyAutoGUI's failsafe check as appeyor moves the mouse???

michaelDCurran committed:
Merge branch 'systemTestWithRobot' into appveyorTests

michaelDCurran committed:
initial.robot: break robot to temporarily bipass freeze.

michaelDCurran committed:
appveyor.yaml: run robot in powershell so exit code is ignored for now.

michaelDCurran committed:
appveyor.yaml:  try again to make system tests return error code 0

michaelDCurran committed:
appveyor.yaml:  try again to make system tests return error code 0

michaelDCurran committed:
appveyor.yaml: fix syntax of uplodFile.

michaelDCurran committed:
system tests: tell appVeyor they are junit formatted?

michaelDCurran committed:
appVeyor: still upload test results on failed builds.

michaelDCurran committed:
Upload unitTest reports to appViewer as well.

michaelDCurran committed:
Investigate freeze in robot test.

michaelDCurran committed:
add some logging to nvdaRobotLib

michaelDCurran committed:
More logging

michaelDCurran committed:
more logging again

michaelDCurran committed:
sleep before connecting to server.

michaelDCurran committed:
more and more logging

michaelDCurran committed:
Fix logging...

michaelDCurran committed:
reimplement quit nvda test inside initial.robot.

michaelDCurran committed:
A try at moving everying to robot language to fix freezes.

michaelDCurran committed:
Launch NVDA from robot with py for appveyor.

michaelDCurran committed:
copy the globalPlugin.

michaelDCurran committed:
Temporarily start rdp for appveyor build

michaelDCurran committed:
bump

michaelDCurran committed:
bump 2

michaelDCurran committed:
bump 3

michaelDCurran committed:
bump 4

michaelDCurran committed:
Slightly re-write now with setup and teardown, and insert q for quit check  and check foreground name is Exit NVDA etc.

michaelDCurran committed:
nvdaRobotLib: use a specific NVDA profile dir.

michaelDCurran committed:
NVDA Spy for robot must use a daemon thread for now to stop  further NVDAs crashing.

michaelDCurran committed:
Robot tests set NVDA not to show the welcome dialog

feerrenrut committed:
allow asserting on last speech

feerrenrut committed:
Merge remote-tracking branch 'origin/robotOnAppveyor' into systemTestWithRobot

feerrenrut committed:
Remove extra tab characters

feerrenrut committed:
Copy in the testSpy globalPlugin for each test.

feerrenrut committed:
Remove sleep

The sleep was required because importing the remote library suceeds
regardless of whether it is possible to execute a keyword. We now test
the connection before we load the library.

feerrenrut committed:
Some cleanup of connection/disconnection code

feerrenrut committed:
Extract code that can be reused and add better documentation

feerrenrut committed:
cleanup systemTest dir

Allow use of different settings files and some cleanup of system test directories

feerrenrut committed:
Assert text on welcome dialog passes

feerrenrut committed:
Copy nvda log for each test

feerrenrut committed:
Try to get systemtest artefacts from appveyor

feerrenrut committed:
Try manually pushing artifacts

feerrenrut committed:
Try recursing over files and pushing manually

feerrenrut committed:
Fix issue when nvdaProfile folders do not exist

michaelDCurran committed:
system tests: don't show usage stats dialog in NVDA

michaelDCurran committed:
appveyor: install NVDA on the build system, and run the system tests against the installed copy.

michaelDCurran committed:
appveyor: install nvda with start-process and wait up to 3 minutes for it to complete, failing on timeout.

michaelDCurran committed:
fix appveyor errors

michaelDCurran committed:
Temprarily include robotremoteserver in NVDA for system tests. Must figure out a better solution.

michaelDCurran committed:
Breakup up appveyor test code into separate powershell code chunks to better show build progress.

michaelDCurran committed:
system tests: use silence synth driver.

michaelDCurran committed:
Merge master

michaelDCurran committed:
system test for exit dialog: just sleep instead of varifying dialog.

michaelDCurran committed:
system tests: don't wait for exit dialog, rather just confirm that the NVDA process definitlly exited.

michaelDCurran committed:
appveyor: move NVDA installation to before_test section.

michaelDCurran committed:
ensure quit system test: exit from the NVDA menu.

michaelDCurran committed:
system tests: try alt tabbing to the exit dialog???

michaelDCurran committed:
More debugging for system tests.

michaelDCurran committed:
IAccessibleHandler: more debugging.

michaelDCurran committed:
IAccessibleHandler: more debugging.

michaelDCurran committed:
eventHandler: more debugging.

michaelDCurran committed:
Up max foreground defer count to 10.

michaelDCurran committed:
Up foreground defers to 100 and log them.

michaelDCurran committed:
IAccessibleHandler: drop foreground defers back down and  allow welcome dialog in  quit from keyboard system test.

michaelDCurran committed:
system tests for quit with keyboard: watch for and close welcome dialog first.

michaelDCurran committed:
quit with keyboard system test: wait longer for NVDA to exit.

michaelDCurran committed:
Restore IAccessibleHandler and eventHandler.

feerrenrut committed:
Not able to get an index returned from get_index_of_speech

feerrenrut committed:
tests passing

feerrenrut committed:
move spy lib logic into spy

feerrenrut committed:
Install systemTestSpy as a package on setup

feerrenrut committed:
Cleanup files, add copyright headers

feerrenrut committed:
Use _ to prevent imports from leaking as keywords

feerrenrut committed:
Move non NVDA specific helpers into a new library

feerrenrut committed:
Merge remote-tracking branch 'origin/installed_systemTestWithRobot' into systemTestWithRobot

feerrenrut committed:
Create self contained package for system test spy

The system test spy now copies in library dependencies that are not
provided by a NVDA install. This is so that the system tests can be run
on appVeyor against an installed version of NVDA

feerrenrut committed:
Zip system test output before pushing to artifacts

feerrenrut committed:
try adding robotlibraries directory to the path

feerrenrut committed:
Fix case of file name

This should resolve the appveyor system test failures

feerrenrut committed:
further tidy up

feerrenrut committed:
Introduce suite name

feerrenrut committed:
fix appveyor system tests

testOutput directories need to be created before writting files to them!

feerrenrut committed:
Stop double nvda install log artifact

feerrenrut committed:
set robot library path

feerrenrut committed:
reduce suite verbosity on appveyor

feerrenrut committed:
Revert to running from repo root

Running from tests/system/ did not reduce the number of suites mentioned in Robots reports

feerrenrut committed:
Clean up some of the appveyor code

leonardder commented on appveyor.yml:
```diff
@@ -109,6 +111,58 @@ build_script:
 
+before_test:
+ - py -m pip install robotframework
+ - py -m pip install robotremoteserver
```
Is there a particular reason why this installed twice, also after running submodule update?

feerrenrut replied:
Perhaps a mistake, for future reference, the other is on [line 65 ](https://github.com/nvaccess/nvda/blob/25f18979960dba1dd1f86b3cf4b34fcacc906b38/appveyor.yml#L65)

feerrenrut replied:
Ah yes, I remember now. For a little while we were packaging robotRemoteServer with the install of NVDA so that it could be used by the addon. Now we package it within the addon. I will remove this line.

leonardder commented on appveyor.yml:
```diff
@@ -109,6 +111,58 @@ build_script:
+ - py -m pip install robotframework
+ - py -m pip install robotremoteserver
+ - py -m pip install pyautogui
+ - py -m pip install nose
```
You can also install them in one run, I'd imagine that could be slightly faster.

feerrenrut replied:
Yes perhaps so, I'm not sure how much time this will save. It might even make sense to convert this to a requirements file in source control. It would make it easier for developers to ready their own systems to run the system tests.

leonardder commented on source/core.py:
```diff
@@ -29,6 +29,11 @@
+import extensionPoints
+
+# inform those who want to know that NVDA has finished starting up.
+postNvdaStartup = extensionPoints.Action()
```
I believe that, according to earlier convention discussions we had, this should be post_NVDAStartup. Also note that it might make sense to write NVDA in upper case, as in NVDAObjects, NVDAHelper

leonardder commented on source/speech.py:
```diff
@@ -27,6 +27,10 @@
+import extensionPoints
+
+# inform those who want to know that there is new speech
+preSpeech = extensionPoints.Action()
```
Also, this might have to be pre_speech according to the conventions.

feerrenrut replied:
I think the conventions were discussed on PR #7606, though the underscores were dropped from that diff. I would be happy with the underscores and certainly want to keep it consistent, perhaps it should be made official on the wiki so that everyone is on the same page. @josephsl  would you be on board with this?

feerrenrut replied:
There may be a minor performance impact here, very little compared to the rest of what happens before speech goes to the synth.

Which raises a bigger question about the approach of using  an extension point to gather the speech, should we be capturing the behaviour that happens between this line, and the eventual `getSynth().speak(speechSequence)`?

I think we probably should, otherwise there is no way for us to test things like auto language / dialect switching or symbol levels.

Perhaps, instead of an extension point for speech, the addon should install and select a dummy synthesizer.

derekriemer replied:
both could be useful actually. The synth captures the output and the extension point the first call.

leonardder commented on tests/system/NVDA Core/variables.py:
```diff
@@ -0,0 +1,14 @@
+WELCOME_DIALOG_TEXT = (
```
This might break if a user runs the system tests with a system language other than English, in which the user default language differs.

feerrenrut replied:
Yes, it's likely that there will be lots of ways that the tests will not pass on various different systems. To start with, we will be using AppVeyor as the target. Over time, if people are trying to run the tests locally, then the tests will become more tolerant of variations in the underlying system.

LeonarddeR reviewed (commented):
Congratulations with this huge work. Note that I haven't yet looked into the system testing framework itself, as it looks a bit complex and I'd have to dive into the matter a little more. Some initial remarks though.

feerrenrut committed:
clean up some names and docs for tests

feerrenrut committed:
Add scons command to run system tests

feerrenrut committed:
ensure dir is created for std out/err logs

Brian1Gaff commented:
Ah I see so keep on reporting funnies then.
 Brian

josephsl commented:
Hi,

Based on what I wrote above, I propose keeping it camel case for now until we get consensus from other stakeholders that we should change this. The chief reason for this is in order to always remember that community is important. Personally, I'll follow the majority opinion if it benefits not only devleopers, but also others.

Thanks.

derekriemer commented on tests/system/libraries/systemTestSpy.py:
```diff
@@ -0,0 +1,207 @@
+import threading
+from systemTestUtils import _blockUntilConditionMet
+from logHandler import log
+from time import clock as _timer
```
why does this name get changed to _timer instead of _clock

feerrenrut replied:
No real reason, except that originally I was importing `from timeit import default_timer as _timer`

derekriemer commented on tests/system/libraries/systemTestSpy.py:
```diff
@@ -0,0 +1,207 @@
+
+	# Private helper methods
+	def _flattenCommandsSeparatingWithNewline(self, commandArray):
+		f = [c for commands in commandArray for newlineJoined in [commands, [u"\n"]] for c in newlineJoined]
```
that's a work of art.

derekriemer commented on tests/system/libraries/systemTestSpy.py:
```diff
@@ -0,0 +1,207 @@
+	def _getJoinedBaseStringsFromCommands(self, speechCommandArray):
+		wsChars = whitespaceMinusSlashN
+		baseStrings = [c.strip(wsChars) for c in speechCommandArray if isinstance(c, basestring)]
+		return ''.join(baseStrings).strip()
```
Why does this need restripped?

feerrenrut replied:
It possibly doesn't need to be there. There is the possibility of a trailing/leading newline character, we want to preserve newline characters within the string. But not at the ends, since they are harder to notice, and can make defining the expected text a frustrating experience.

derekriemer commented on tests/system/libraries/systemTestSpy.py:
```diff
@@ -0,0 +1,207 @@
+		return self._spy.getSpeechSinceIndex(speechIndex)
+
+	def reset_all_speech_index(self):
+		self._allSpeechStartIndex = self._spy.getIndexOfLastSpeech()
```
The name suggests this should reset the speech. Does this actually do anything but return the last index?

feerrenrut replied:
Hmm, perhaps it's because its not clear that "all speech" is a grouping. This resets the "all speech" index, meaning that when you next call `get_all_speech` it will return any new speech added after the index was reset. It's important that we don't actually clear the internal buffer, since tests may need to hold on to speech indexes and be able to refer to them later, without them being invalidated.

derekriemer reviewed (commented):
This is very cool. a few comments.

feerrenrut commented on appveyor.yml:
```diff
@@ -109,6 +111,58 @@ build_script:
+     if(!$env:release) {
+      $nvdaLauncherFile+="_snapshot"
+     }
+     $nvdaLauncherFile+="_${env:version}.exe"
```
Yes, the tests run on try builds, against an install of NVDA. I can't speak to your second question, it might be best for @michaelDCurran to answer that one.

feerrenrut commented on tests/sconscript:
```diff
@@ -20,6 +20,11 @@ unitTests = env.SConscript("unit/sconscript", exports=["env"])
 
+systemTests = env.SConscript("system/sconscript", exports=["env"])
+env.Depends(systemTests, sourceDir)
+env.AlwaysBuild(systemTests)
```
At the moment, the system tests only run if you explicitly give the option when starting scons. Then we want the tests to run, even if there has been no changes to dependencies.

feerrenrut commented on tests/system/libraries/nvdaRobotLib.py:
```diff
@@ -0,0 +1,219 @@
+		_pJoin(systemTestSourceDir, "libraries", systemTestSpyAddonName+".py"),
+		_pJoin(profileSysTestSpyPackageStagingDir, "__init__.py")
+	)
+	opSys.copy_file(
```
Within the system tests I have preferred to use methods available in the system test framework, since they tend to provide more robustness, checks, and reporting when something is not right.  This makes it easier to track down the problem in the system test log.  Even when things go correctly, there are entries (including timestamps) added to the system test log which can be quite helpful.

feerrenrut commented:
@josephsl  and @leonardder re: underscore vs camel-case 

@josephsl I agree with you that source code is a communicative object, and that it's written for people not machines (otherwise we would all be writing in some form of machine code). Code is read many more times than it is written, and as such, being clear is important. 

My vote would be on using underscore, because it more clearly defines the pattern. While aiming for consistency across the code base is important, consistency for a paradigm is more so. I couldn't find any opposition to the use of underscores in #7606 where we have already had an extended discussion of this topic. So as the next action, I suggest we all convert our various extension point introducing PR's to use underscores, and I will add a wiki page to document the decision and rationale for the naming scheme.

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into systemTestWithRobot

feerrenrut commented:
@josephsl https://github.com/nvaccess/nvda/wiki/Contributing#identifier-names has now been edited.

josephsl commented:
Hi, applying to #7606 now. Thanks.

 

From: Reef Turner <notifications@github.com> 
Sent: Thursday, July 5, 2018 10:54 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] System test with robot framework (#8481)

 

@josephsl <https://github.com/josephsl>  https://github.com/nvaccess/nvda/wiki/Contributing#identifier-names has now been edited.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8481#issuecomment-402933678> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkOATEScyqUYO0mGk8kaOsBLl_QY8ks5uDvuJgaJpZM4VDWfR> .

feerrenrut committed:
Use dummy speech synth to get speech

This captures all NVDA manipulation of speech before it is given to
synthesizers.

feerrenrut committed:
wait longer before sending input

feerrenrut committed:
Stop update dialog from showing on startup

feerrenrut committed:
simplify the appveyor pip installs

michaelDCurran committed:
Update What's new
